diff --git a/auto_round/calib_dataset.py b/auto_round/calib_dataset.py
index 9910b8d2..bad3fe90 100644
--- a/auto_round/calib_dataset.py
+++ b/auto_round/calib_dataset.py
@@ -432,6 +432,7 @@ def get_opencode_instruct_dataset(
     os_cnt, have_bos, have_eos = 0, False, False
 
     for input_id in input_ids:
+        os_cnt, have_bos, have_eos = 0, False, False
         if input_id[0] == bos_token_id:
             input_id = input_id[1:]
             os_cnt, have_bos = os_cnt + 1, True
@@ -926,6 +927,7 @@ def _get_dataset_impl(tokenizer, seqlen, dataset_name="NeelNanda/pile-10k", seed
         os_cnt, have_bos, have_eos = 0, False, False
 
         for input_id in input_ids:
+            os_cnt, have_bos, have_eos = 0, False, False
             if input_id[0] == bos_token_id:
                 input_id = input_id[1:]
                 os_cnt, have_bos = os_cnt + 1, True
diff --git a/auto_round/wrapper.py b/auto_round/wrapper.py
index f36b67e0..273e57c6 100644
--- a/auto_round/wrapper.py
+++ b/auto_round/wrapper.py
@@ -136,6 +136,14 @@ class WrapperLinear(torch.nn.Module):
     def bias(self):
         return self.orig_layer.bias
 
+    @property
+    def in_features(self):
+        return self.orig_layer.in_features
+
+    @property
+    def out_features(self):
+        return self.orig_layer.out_features
+
     def _init_tuning_params_and_quant_func(self):
         """Initializes tuning parameters and quantization functions.
 
